### DISCUSSION TITLE
Custom scalars

### DISCUSSION THREAD

**Discussion by mattfysh** [urql-graphql/urql#1785] [Q&A] [answered]
Hi there, I'm working on an AWS Amplify project and some of the fields have custom scalars. The one I'm working with is `AWSJSON` and I'm hoping to understand the best way to implement centralized serialization and deserialization somewhere in urql or its exchanges.

At the moment, I'm using graphcache with custom resolvers. If you imagine a schema such as:

```graphql
type TypeA {
  customField: AWSJSON
}
type TypeB {
  customField: AWSJSON
 }
input TypeInput {
  customInput: AWSJSON
}
```

My custom resolvers currently allow for deserializing the response from the GraphQL server:

```js
resolvers: {
  TypeA: {
    customField: typeA => JSON.parse(typeA.customField),
  },
  TypeB: {
    customField: typeB => JSON.parse(typeB.customField),
  }
}
```

Two questions:

1. You can see the duplication of the deserializer, and the need to repeat this as many times as the custom scalar appears in the query schemas. How can I specify a global deserializer for `AWSJSON`, which is invoked anywhere the custom scalar appears in the query schemas?
2. How can I also implement a serializer for input types, i.e. any AWSJSON custom scalar fields that are used as inputs in the mutation schemas?

**Comment by kitten** [comment, marked as answer]
> My custom resolvers currently allow for deserializing the response from the GraphQL server:

Something's fishy here! Your API returns a GraphQL result as JSON and your JSON type is still serialised as a string in that JSON. So in other words, you're accidentally using this type on AppSync not to send JSON but to send JSON strings? It sounds like there's some kind of double serialisation going in, so definitely just fix that in your AppSync code, if you can. That sounds like a mistake 😅

> How can I specify a global deserializer for AWSJSON, which is invoked anywhere the custom scalar appears in the query schemas

Yes, there's a third-party exchange that avoids having to configure Graphcache for scalars. While we'd say there are some caveats for transforming data at the exchange level, 1:1 transformations are fine: https://www.npmjs.com/package/urql-custom-scalars-exchange

But again, it sounds like this serialisation/deserialisation in particular shouldn't be necessary; rather, it sounds like something is stringifying JSON on the AppSync side, which is of course then still a string and hence can be serialised as JSON, but it's not the raw JSON data.

**Reply by mattfysh** [reply]
The scalar type `AWSJSON` is a JSON-serialized string, as implemented by AWS AppSync:

> A JSON string. Any valid JSON construct is automatically parsed and loaded in the resolver mapping templates as maps, lists, or scalar values rather than as the literal input strings. Unquoted strings or otherwise invalid JSON result in a GraphQL validation error.

https://docs.aws.amazon.com/appsync/latest/devguide/scalars.html

re: attaching resolvers to types (in addition to fields), see Trivial Resolvers & Scalar Coercion in the GraphQL spec: https://graphql.org/learn/execution/ • They also created a blog post where this approach is used for mocking test data: https://graphql.org/blog/mocking-with-graphql/

```
// mocking data by implementing resolvers on types (in contrast to field-level resolvers)
{
  Int: () => 6,
  Float: () => 22.1,
  String: () => 'Hello',
}
```

It would be pretty cool to add this part of the GraphQL spec to the "local resolvers" feature of graphcache.

## Part 2 - input types serialization

By extension, if the same scalar type is used in the mutation schemas, it needs a serializer

**Reply by kitten** [reply]
I'd be very surprised if that JSON data would still be serialised as a string outside of this context. I've used this before and accepting responses where a raw JSON result was delivered, even inside scalars, was working fine enough using `$util.parseJson(String)`.

In general, we're not currently interested in adding this to Graphcache as it adds additional complexity in code generation and the `resolvers` signature just for the scalar case. It also needs schema information to work, since otherwise recognising scalars is impossible.

It's also a feature that's rarely needed for UIs to actually use such scalars, given the guarantees already present, i.e. there aren't many cases where this coercion actually makes sense. Typically, JSON would be represented as a `JSON` scalar that isn't double-serialised, dates for instance would be further formatted, so it's not common to parse them and then do a human-readable stringification again separately, since both can be done in one go in UI components, etc etc.

But again, there is that third party exchange that takes care of this, which then requires the schema information which can be generated using GraphQL Code Generator's integration to `@urql/introspection` or `@urql/introspection` itself. https://www.npmjs.com/package/urql-custom-scalars-exchange
